pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/MultiStageStatsTreeBuilder.java

```diff
@@ -19,24 +19,46 @@
 package org.apache.pinot.query.runtime;
 
 import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.ArrayList;
 import java.util.List;
+import org.apache.pinot.query.planner.physical.DispatchablePlanFragment;
 import org.apache.pinot.query.planner.plannode.PlanNode;
 import org.apache.pinot.query.runtime.plan.MultiStageQueryStats;
+import org.apache.pinot.spi.utils.JsonUtils;
 
 
 public class MultiStageStatsTreeBuilder {
   private final List<PlanNode> _planNodes;
   private final List<? extends MultiStageQueryStats.StageStats> _queryStats;
+  private final List<DispatchablePlanFragment> _planFragments;
 
-  public MultiStageStatsTreeBuilder(List<PlanNode> planNodes,
+  public MultiStageStatsTreeBuilder(List<DispatchablePlanFragment> planFragments,
       List<? extends MultiStageQueryStats.StageStats> queryStats) {
-    _planNodes = planNodes;
+    _planFragments = planFragments;
+    _planNodes = new ArrayList<>(planFragments.size());
+    for (DispatchablePlanFragment stagePlan : planFragments) {
+      _planNodes.add(stagePlan.getPlanFragment().getFragmentRoot());
+    }
     _queryStats = queryStats;
   }
 
   public ObjectNode jsonStatsByStage(int stage) {
-    MultiStageQueryStats.StageStats stageStats = _queryStats.get(stage);
     PlanNode planNode = _planNodes.get(stage);
+
+    MultiStageQueryStats.StageStats stageStats = stage < _queryStats.size() ? _queryStats.get(stage) : null;
+    if (stageStats == null) {
+      // We don't have stats for this stage. This can happen when the stage is not executed. For example, when there
+      // are no segments for a table.
+      ObjectNode jsonNodes = JsonUtils.newObjectNode();
+      jsonNodes.put("type", "EMPTY_MAILBOX_SEND");
+      jsonNodes.put("stage", stage);
+      jsonNodes.put("description", "No stats available for this stage. It may have been pruned.");
+      String tableName = _planFragments.get(stage).getTableName();
+      if (tableName != null) {
+        jsonNodes.put("table", tableName);
+      }
+      return jsonNodes;
+    }
```

Comment:
Just for my understanding, this should only ever be possible for leaf stages right?

Reply:
From this class perspective, it could happen with any stage that is not executed.

From my understanding, the only stages we generate that end up not being executed is the leaf stages when there are no segments.

If my understanding is incorrect or in the future there are other ways to prune a stage, this code should still work.